hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ABlockOutputStream.java

```diff
@@ -876,11 +877,17 @@ private void uploadBlockAsync(final S3ADataBlocks.DataBlock block,
             ? RequestBody.fromFile(uploadData.getFile())
             : RequestBody.fromInputStream(uploadData.getUploadStream(), size);
 
-        request = writeOperationHelper.newUploadPartRequestBuilder(
+        UploadPartRequest.Builder requestBuilder = writeOperationHelper.newUploadPartRequestBuilder(
             key,
             uploadId,
             currentPartNumber,
-            size).build();
+            size);
+
+        if (isLast) {
```

Comment:
add this as a param to pass down to WriteOperationHelper and let the factory do the work